lib/std/http/Server.zig

```diff
@@ -329,21 +352,61 @@ pub const Response = struct {
 
     transfer_encoding: ResponseTransfer = .none,
 
-    server: *Server,
+    allocator: Allocator,
     address: net.Address,
     connection: BufferedConnection,
 
     headers: http.Headers,
     request: Request,
 
+    state: State = .first,
+
+    const State = enum {
+        first,
+        start,
+        waited,
+        responded,
+        finished,
+    };
+
     pub fn deinit(res: *Response) void {
-        res.server.allocator.destroy(res);
+        res.connection.close();
+
+        res.headers.deinit();
+        res.request.headers.deinit();
+
+        if (res.request.parser.header_bytes_owned) {
+            res.request.parser.header_bytes.deinit(res.allocator);
+        }
     }
 
+    pub const ResetState = enum { reset, closing };
+
     /// Reset this response to its initial state. This must be called before handling a second request on the same connection.
-    pub fn reset(res: *Response) void {
-        res.request.headers.deinit();
-        res.headers.deinit();
+    pub fn reset(res: *Response) ResetState {
+        if (res.state == .first) {
+            res.state = .start;
+            return .reset;
+        }
+
+        if (!res.request.parser.done) {
+            // If the response wasn't fully read, then we need to close the connection.
+            res.connection.conn.closing = true;
+            return .closing;
+        }
+
+        // A connection is only keep-alive if the Connection header is present and it's value is not "close".
+        // The server and client must both agree
+        const res_connection = res.headers.getFirstValue("connection");
+        const res_keepalive = res_connection != null and !std.ascii.eqlIgnoreCase("close", res_connection.?);
+
+        const req_connection = res.request.headers.getFirstValue("connection");
+        const req_keepalive = req_connection != null and !std.ascii.eqlIgnoreCase("close", req_connection.?);
+        if (res_keepalive and req_keepalive) {
+            res.connection.conn.closing = false;
+        } else {
+            res.connection.conn.closing = true;
+        }
```

Comment:
:( 
Still not quite correct ...

I have had to change line 401 to this to get keep alive working

const res_keepalive = res_connection == null or !std.ascii.eqlIgnoreCase("close", res_connection.?);

The server will report connection = keep-alive because of the default case in res.do(), but the res.header("connection") is never getting set, so it ends up returning .closing() on this reset.

So you end up with the server reporting keep-alive over the wire during res.do(), but then setting res.connection.conn.closing = true on the next call to res.reset(), and returning the value .closing to the callsite of res.reset(). 

Suggest that in res.wait() .. after its parsed the request headers, that it sets res.headers("connection") = res.request.header("connection"), to avoid any confusion.  

Either that, or toggle the logic here to treat a missing res.headers("connection") as defaulting to keep-alive.  (Its more efficient to do toggle the logic here, rather than add to the arraylist, but its not obvious to understand)